projects/packages/waf/src/class-waf-rules-manager.php

```diff
@@ -0,0 +1,321 @@
+		/**
+		 * WordPress filesystem abstraction.
+		 *
+		 * @var \WP_Filesystem_Base $wp_filesystem
+		 */
+		global $wp_filesystem;
+
+		Waf_Runner::initialize_filesystem();
+
+		$rules                = "<?php\n";
+		$entrypoint_file_path = Waf_Runner::get_waf_file_path( self::RULES_ENTRYPOINT_FILE );
+
+		// Ensure that the folder exists
+		if ( ! $wp_filesystem->is_dir( dirname( $entrypoint_file_path ) ) ) {
+			$wp_filesystem->mkdir( dirname( $entrypoint_file_path ) );
+		}
+
+		// Ensure all potentially required rule files exist
+		$rule_files = array( self::RULES_ENTRYPOINT_FILE, self::AUTOMATIC_RULES_FILE, self::IP_ALLOW_RULES_FILE, self::IP_BLOCK_RULES_FILE );
+		foreach ( $rule_files as $rule_file ) {
+			$rule_file = Waf_Runner::get_waf_file_path( $rule_file );
+			if ( ! $wp_filesystem->is_file( $rule_file ) ) {
+				if ( ! $wp_filesystem->put_contents( $rule_file, "<?php\n" ) ) {
+					throw new \Exception( 'Failed writing rules file to: ' . $rule_file );
+				}
+			}
+		}
+
+		// Add manual rules
+		if ( get_option( self::IP_LISTS_ENABLED_OPTION_NAME ) ) {
+			$rules .= self::wrap_require( Waf_Runner::get_waf_file_path( self::IP_ALLOW_RULES_FILE ) ) . "\n";
+			$rules .= self::wrap_require( Waf_Runner::get_waf_file_path( self::IP_BLOCK_RULES_FILE ), "return \$waf->block( 'block', -1, 'ip block list' );" ) . "\n";
+		}
+
+		// Add automatic rules
+		if ( get_option( self::AUTOMATIC_RULES_ENABLED_OPTION_NAME ) ) {
+			$rules .= self::wrap_require( Waf_Runner::get_waf_file_path( self::AUTOMATIC_RULES_FILE ) ) . "\n";
+		}
+
+		// Update the rules file
+		if ( ! $wp_filesystem->put_contents( $entrypoint_file_path, $rules ) ) {
+			throw new \Exception( 'Failed writing rules file to: ' . $entrypoint_file_path );
+		}
+	}
+
+	/**
+	 * Generates the automatic-rules.php script
+	 *
+	 * @throws \Exception If rules cannot be generated and saved.
+	 * @return void
+	 */
+	public static function generate_automatic_rules() {
+		/**
+		 * WordPress filesystem abstraction.
+		 *
+		 * @var \WP_Filesystem_Base $wp_filesystem
+		 */
+		global $wp_filesystem;
+
+		Waf_Runner::initialize_filesystem();
+
+		$automatic_rules_file_path = Waf_Runner::get_waf_file_path( self::AUTOMATIC_RULES_FILE );
+
+		// Ensure that the folder exists.
+		if ( ! $wp_filesystem->is_dir( dirname( $automatic_rules_file_path ) ) ) {
+			$wp_filesystem->mkdir( dirname( $automatic_rules_file_path ) );
+		}
+
+		try {
+			$rules = self::get_rules_from_api();
+		} catch ( \Exception $exception ) {
+			// Do not throw API exceptions for users who do not have access
+			if ( 401 !== $exception->getCode() ) {
+				throw $exception;
+			}
+		}
+
+		// If there are no rules available, don't overwrite the existing file.
+		if ( empty( $rules ) ) {
+			return;
+		}
+
+		if ( ! $wp_filesystem->put_contents( $automatic_rules_file_path, $rules ) ) {
+			throw new \Exception( 'Failed writing automatic rules file to: ' . $automatic_rules_file_path );
```

Comment:
@nateweller I keep getting the following Fatal error in my development environment since this was merged:

```
[01-Feb-2023 07:52:03 UTC] PHP Fatal error:  Uncaught Exception: Failed writing automatic rules file to: /var/www/html/wp-content/jetpack-waf/rules/automatic-rules.php in /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-rules-manager.php:242
Stack trace:
#0 /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-rules-manager.php(78): Automattic\Jetpack\Waf\Waf_Rules_Manager::generate_automatic_rules()
#1 /var/www/html/wp-includes/class-wp-hook.php(308): Automattic\Jetpack\Waf\Waf_Rules_Manager::update_rules_cron()
#2 /var/www/html/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters()
#3 /var/www/html/wp-includes/plugin.php(565): WP_Hook->do_action()
#4 /var/www/html/wp-cron.php(188): do_action_ref_array()
#5 {main}
  thrown in /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-rules-manager.php on line 242
```

Anything we can do about that?

Reply:
You bet, I will address this and open up a PR :+1:

Reply:
@jeherve I may have identified the potential cause of this, can you fill me in on the details of your local environment when this fatal is occurring?
* Is the `waf` module currently enabled or disabled?
* Does the `wp-content/jetpack-waf` directory exist? If so, what files are currently in that directory?

Reply:
Opened a PR to address this: https://github.com/Automattic/jetpack/pull/28760